server/s3_copy_object_action.cc

```diff
@@ -0,0 +1,208 @@
+    std::shared_ptr<S3MotrReaderFactory> motrreader_s3_factory,
+    std::shared_ptr<S3MotrKVSWriterFactory> kv_writer_factory)
+    : S3ObjectAction(std::move(req), std::move(bucket_meta_factory),
+                     std::move(object_meta_factory)),
+      write_in_progress(false),
+      read_in_progress(false) {
+  s3_log(S3_LOG_DEBUG, request_id, "Constructor\n");
+  s3_log(S3_LOG_INFO, request_id,
+         "S3 API: CopyObject. Destination: [%s], Source: [%s]\n",
+         request->get_object_uri().c_str(),
+         request->get_copy_object_source().c_str());
+  s3_copy_action_state = S3CopyObjectActionState::empty;
+
+  old_object_oid = {0ULL, 0ULL};
+  old_layout_id = -1;
+  new_object_oid = {0ULL, 0ULL};
+
+  if (motr_api) {
+    s3_motr_api = std::move(motr_api);
+  } else {
+    s3_motr_api = std::make_shared<ConcreteMotrAPI>();
+  }
+  S3UriToMotrOID(s3_motr_api, request->get_object_uri().c_str(), request_id,
+                 &new_object_oid);
+  // Note valid value is set during create object
+  layout_id = -1;
+  tried_count = 0;
+  salt = "uri_salt_";
+
+  if (motrwriter_s3_factory) {
+    motr_writer_factory = std::move(motrwriter_s3_factory);
+  } else {
+    motr_writer_factory = std::make_shared<S3MotrWriterFactory>();
+  }
+  if (motrreader_s3_factory) {
+    motr_reader_factory = std::move(motrreader_s3_factory);
+  } else {
+    motr_reader_factory = std::make_shared<S3MotrReaderFactory>();
+  }
+  setup_steps();
+};
+
+void S3CopyObjectAction::setup_steps() {
+  s3_log(S3_LOG_DEBUG, request_id, "Setting up the action\n");
+  ACTION_TASK_ADD(S3CopyObjectAction::validate_copyobject_request, this);
+  ACTION_TASK_ADD(S3CopyObjectAction::create_object, this);
+  ACTION_TASK_ADD(S3CopyObjectAction::read_object, this);
+  ACTION_TASK_ADD(S3CopyObjectAction::initiate_data_streaming, this);
+  ACTION_TASK_ADD(S3CopyObjectAction::save_metadata, this);
+  ACTION_TASK_ADD(S3CopyObjectAction::send_response_to_s3_client, this);
+}
+
+// Destination bucket
+void S3CopyObjectAction::fetch_bucket_info_failed() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  s3_copy_action_state = S3CopyObjectActionState::validationFailed;
+  if (bucket_metadata->get_state() == S3BucketMetadataState::missing) {
+    s3_log(S3_LOG_DEBUG, request_id, "Bucket not found\n");
+    set_s3_error("NoSuchBucket");
+  } else if (bucket_metadata->get_state() ==
+             S3BucketMetadataState::failed_to_launch) {
+    s3_log(S3_LOG_ERROR, request_id,
+           "Bucket metadata load operation failed due to pre launch failure\n");
+    set_s3_error("ServiceUnavailable");
+  } else {
+    s3_log(S3_LOG_DEBUG, request_id, "Bucket metadata fetch failed\n");
+    set_s3_error("InternalError");
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+}
+
+// Destination object
+void S3CopyObjectAction::fetch_object_info_failed() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  next();
```

Comment:
I believe, fetch_object_info_failed() is in the context of 'source' object.
When source object is not found, we may have to return to client. Check how AWS S3 behaves, and then implement this.

Reply:
No Datta, the source object and bucket, are sent as part of a request header, and our RequestObject class member variables (object and bucket) are initialized with bucket and object in URI (which will be destination bucket and object)

This callback function is same as written in S3PutObjectAction class. 

We still have to handle validation on source bucket and object, in next PR.

Reply:
ok